h5netcdf/core.py

```diff
@@ -278,74 +279,68 @@ def __len__(self):
     @property
     def dtype(self):
         """Return NumPy dtype object giving the variable’s type."""
-        return self._h5ds.dtype
-
-    def _get_padding(self, key):
-        """Return padding if needed, defaults to False."""
-        padding = False
-        if self.dtype != str and self.dtype.kind in ["f", "i", "u"]:
-            key0 = _expanded_indexer(key, self.ndim)
-            key0 = _transform_1d_boolean_indexers(key0)
-            # extract max shape of key vs hdf5-shape
-            h5ds_shape = self._h5ds.shape
-            shape = self.shape
-
-            # check for ndarray and list
-            # see https://github.com/pydata/xarray/issues/7154
-            # first get maximum index
-            max_index = [
-                max(k) + 1 if isinstance(k, (np.ndarray, list)) else k.stop
-                for k in key0
-            ]
-            # second convert to max shape
-            max_shape = tuple(
-                [
-                    shape[i] if k is None else max(h5ds_shape[i], k)
-                    for i, k in enumerate(max_index)
-                ]
-            )
-
-            # check if hdf5 dataset dimensions are smaller than
-            # their respective netcdf dimensions
-            sdiff = [d0 - d1 for d0, d1 in zip(max_shape, h5ds_shape)]
-            # create padding only if hdf5 dataset is smaller than netcdf dimension
-            if sum(sdiff):
-                padding = [(0, s) for s in sdiff]
-        return padding
+        if self._dtype is None:
+            self._dtype = self._h5ds.dtype
+        return self._dtype
 
     def __array__(self, *args, **kwargs):
         return self._h5ds.__array__(*args, **kwargs)
 
-    def __getitem__(self, key):
-        from .legacyapi import Dataset
+    h5py_version = version.parse(h5py.__version__)
+    if version.parse("3.0.0") <= h5py_version < version.parse("3.7.0"):
+        def __getitem__(self, key):
+            from .legacyapi import Dataset
 
-        if isinstance(self._parent._root, Dataset):
-            # this is only for legacyapi
-            # fix boolean indexing for affected versions
-            # https://github.com/h5py/h5py/pull/2079
-            # https://github.com/h5netcdf/h5netcdf/pull/125/
-            h5py_version = version.parse(h5py.__version__)
-            if version.parse("3.0.0") <= h5py_version < version.parse("3.7.0"):
+            if isinstance(self._parent._root, Dataset):
```

Comment:
I think these accesses to the file structure are somewhat slow. I've just forked off the implementations as two separate ones.